### ISSUE TITLE
Question about training self-supervised model

### ISSUE THREAD

**Issue by 0three** [Philip-Bachman/amdim-public#4]
Hi,

Thanks for your sharing code first. However, the defaults setting of loss_opt seems that it is not a self-supervised task. The loss_opt utilize the information of label obviously in default setting during training phase. 

Is this a problem?

**Comment by 0three** [comment]
I checked the paper. I'm trying to train the model again with loss_inf alone.

**Comment by 0three** [comment]
Success.

**Comment by wbuchwalter** [comment]
Hi,
`loss_cls` is to optimize the linear classifiers to allow to get some feedback on how the model is training.
If you look here: https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L162 you will see that the classifier is detached from the encoder, so `loss_cls` has no impact on the encoder. The encoder is only trained using `loss_inf` which is not using the labels.

**Comment by 0three** [comment]
I still cannot understand the truly meaning of mlp and cls. Maybe I need to read more details in your paper. 
I only need to use this code to do a simple non-label self-supervised task. I just remove the loss_cls of loss_inf and use mlp as the final logit. 
If this is a wrong behaviour, plz correct me.

Thanks for your explanation very much!

**Comment by wbuchwalter** [comment]
If you only want to train the encoder, then yes you can get rid of the evaluator in the model ([here](https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L199) and [here](https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L289) and remove `loss_cls`.
Then, once you are done with self-supervised training, you can use `model.encode(x)` to get the representationsfor a batch of inputs.
But I am not sure what you are hoping to achieve by keeping the mlp classifier like you were saying if you don't have any label as it will never be trained?

----------------------
More details on how AMDIM works for reference:

AMDIM is about self-supervised representation learning, so learning to generate good representations without relying on labels.
These representations are generated by an encoder (that you can find here: [Encoder](https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L16)).
Hence we need to be able to train the encoder without labels, which is what is happening in the logic leading to `loss_inf` (you can find more info in section 3 of our paper).

However, we also need to be able to evaluate how good the learned representations are (to compare our results to other approaches for example). To do this, we have trained 2 classifiers in a supervised way on top of our encoder representations (you can see the two classifiers [here](https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L150)).
One of these classifiers is a simple linear classifier (no hidden layer), the other one is an MLP (multilayer perceptron) with a hidden layer size of 1024. 
These are referred to as `linear` and `MLP` in our paper (e.g, in Table 1 (b) and (c) ), or `lin` and `mlp`  in our code.
`loss_cls` stands for the classifier (cls) loss, and is just the aggregate of the cross-entropy losses of both classifiers ([here](https://github.com/Philip-Bachman/amdim-public/blob/master/task_self_supervised.py#L57)).

To recap: 
* The encoder is optimized via `loss_inf` that does not make use of labels.
* The classifiers are optimized via `loss_cls` in a traditional supervised way, and **do need labels** to be trained
    * The classifiers are detached from the encoder ([here}(https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L162)), as we don't want to backprop the supervised loss' gradients through the encoder.
